pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,145 @@
+            "Doing so raises a RuntimeError.",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            msg_id = None
+            if self._modified_iterating_list_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-list"
+            elif self._modified_iterating_dict_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-dict"
+            elif self._modified_iterating_set_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-set"
+            if msg_id is not None:
+                self.add_message(
+                    msg_id,
+                    node=node,
+                    args=(iter_obj.name,),
+                    confidence=interfaces.INFERENCE,
+                )
+                break  # since the msg is raised for the `for` node no further check is needed
+
+    @staticmethod
+    def _is_node_expr_calls_attribute_name(node: nodes.NodeNG) -> bool:
+        return (
+            isinstance(node, nodes.Expr)
+            and isinstance(node.value, nodes.Call)
+            and isinstance(node.value.func, nodes.Attribute)
+            and isinstance(node.value.func.expr, nodes.Name)
+        )
+
+    @classmethod
+    def _common_cond_list_set(
+        cls,
+        node: nodes.NodeNG,
+        list_obj: nodes.NodeNG,
+        infer_val: Optional[nodes.NodeNG],
+    ) -> bool:
+        return (infer_val == utils.safe_infer(list_obj)) and (
+            node.value.func.expr.name == list_obj.name
+        )
+
+    @staticmethod
+    def _is_node_assigns_subscript_name(node: nodes.NodeNG) -> bool:
+        return isinstance(node, nodes.Assign) and (
+            isinstance(node.targets[0], nodes.Subscript)
+            and (isinstance(node.targets[0].value, nodes.Name))
+        )
+
+    @classmethod
+    def _modified_iterating_list_cond(
+        cls, node: nodes.NodeNG, list_obj: nodes.NodeNG
+    ) -> bool:
+        if not cls._is_node_expr_calls_attribute_name(node):
+            return False
+        infer_val = utils.safe_infer(node.value.func.expr)
+        if not isinstance(infer_val, nodes.List):
+            return False
+        return (
+            cls._common_cond_list_set(node, list_obj, infer_val)
+            and node.value.func.attrname in __LIST_MODIFIER_METHODS__
+        )
+
+    @classmethod
+    def _modified_iterating_dict_cond(
+        cls, node: nodes.NodeNG, list_obj: nodes.NodeNG
+    ) -> bool:
+        if not cls._is_node_assigns_subscript_name(node):
+            return False
+        infer_val = utils.safe_infer(node.targets[0].value)
+        if not isinstance(infer_val, nodes.Dict):
+            return False
```

Comment:
This line is uncovered. Can we make a test case that hits this line?

Reply:
Sure, I found something and added to the test  :)